File: backend/app/models/extraction.py
import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, model_validator


class SourceType(str, Enum):
    SQL_POSTGRESQL = "SQL_POSTGRESQL"
    SQL_MYSQL = "SQL_MYSQL"
    SQL_SQLITE = "SQL_SQLITE"
    JSON = "JSON"


class ErrorCode(str, Enum):
    NO_CONNECTION = "NO_CONNECTION"
    SECURITY_REJECTION = "SECURITY_REJECTION"
    QUERY_SYNTAX = "QUERY_SYNTAX"
    TARGET_NOT_FOUND = "TARGET_NOT_FOUND"
    PERMISSION_DENIED = "PERMISSION_DENIED"
    TIMEOUT = "TIMEOUT"
    AMBIGUOUS_PROMPT = "AMBIGUOUS_PROMPT"
    SOURCE_UNAVAILABLE = "SOURCE_UNAVAILABLE"
    UNKNOWN = "UNKNOWN"


class ColumnDescriptor(BaseModel):
    name: str = Field(min_length=1)
    type: Literal["string", "integer", "float", "boolean", "datetime", "null", "unknown"]


class QueryPlan(BaseModel):
    language: Literal["sql", "jsonpath"]
    expression: str
    parameters: Optional[dict[str, Any]] = None
    generated_by_model: Optional[str] = None


class ExtractionError(BaseModel):
    code: ErrorCode
    message: str
    technical_detail: Optional[str] = None


class DataExtraction(BaseModel):
    contract_version: Literal["v1"] = "v1"
    extraction_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    session_id: str = Field(min_length=1)
    connection_id: str = Field(min_length=1)
    source_type: SourceType
    query_plan: QueryPlan
    columns: list[ColumnDescriptor] = Field(default_factory=list)
    rows: list[dict[str, Any]] = Field(default_factory=list)
    row_count: int = Field(ge=0)
    truncated: bool = False
    status: Literal["success", "error"]
    error: Optional[ExtractionError] = None
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    @model_validator(mode="after")
    def validate_error_presence(self) -> "DataExtraction":
        if self.status == "error" and self.error is None:
            raise ValueError("error must be set when status='error'")
        if self.status == "success" and self.error is not None:
            raise ValueError("error must be None when status='success'")
        return self


class AgentTrace(BaseModel):
    trace_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    extraction_id: str
    pipeline: Literal["sql", "json"]
    query_display: str
    preview_rows: list[dict[str, Any]] = Field(default_factory=list)
    preview_columns: list[ColumnDescriptor] = Field(default_factory=list)
    security_rejection: bool = False
    collapsed: bool = True
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: frontend/src/types/extraction.ts
/**
 * Tipos TypeScript para el contrato `data_extraction.v1`.
 *
 * Mantener alineado 1:1 con `specs/003-data-agent/contracts/data-extraction-v1.schema.json`.
 * Consumido por el chat (Feature 003) y el Widget Generation (Feature 004).
 */

export type SourceType =
  | "SQL_POSTGRESQL"
  | "SQL_MYSQL"
  | "SQL_SQLITE"
  | "JSON";

export type QueryLanguage = "sql" | "jsonpath";

export type ColumnType =
  | "string"
  | "integer"
  | "float"
  | "boolean"
  | "datetime"
  | "null"
  | "unknown";

export type ExtractionStatus = "success" | "error";

export type ErrorCode =
  | "NO_CONNECTION"
  | "SECURITY_REJECTION"
  | "QUERY_SYNTAX"
  | "TARGET_NOT_FOUND"
  | "PERMISSION_DENIED"
  | "TIMEOUT"
  | "AMBIGUOUS_PROMPT"
  | "SOURCE_UNAVAILABLE"
  | "UNKNOWN";

export interface QueryPlan {
  language: QueryLanguage;
  expression: string;
  parameters?: Record<string, unknown>;
  generated_by_model?: string;
}

export interface ColumnDescriptor {
  name: string;
  type: ColumnType;
}

export interface ExtractionError {
  code: ErrorCode;
  message: string;
  technical_detail?: string;
}

export type ExtractionRow = Record<string, unknown>;

export interface DataExtraction {
  contract_version: "v1";
  extraction_id: string;
  session_id: string;
  connection_id: string;
  source_type: SourceType;
  query_plan: QueryPlan;
  columns: ColumnDescriptor[];
  rows: ExtractionRow[];
  row_count: number;
  truncated: boolean;
  status: ExtractionStatus;
  error?: ExtractionError | null;
  generated_at: string;
}

export type TracePipeline = "sql" | "json";

export interface AgentTrace {
  trace_id: string;
  extraction_id: string;
  pipeline: TracePipeline;
  query_display: string;
  preview_rows: ExtractionRow[];
  preview_columns: ColumnDescriptor[];
  security_rejection: boolean;
  collapsed: boolean;
  created_at: string;
}
